### PR TITLE
Add eBPF connection tracking

### PR DESCRIPTION
**Pull request by asymmetric**
Introduces connection tracking via eBPF. This allows scope to get notified of every connection event, without relying on the parsing of `/proc/net`, and therefore:
- improve performance
- increase reliability
### How it works

...
### Limitations:
- [ ] Does not support IPv6
- [ ] Sets `procspied: true` on connections coming from eBPF
- [ ] Reads 5000 bytes of `/proc/net/tcp`, whereas the actual file could be larger

### CODE REVIEW

**Comment by alban** [comment]
You should have a better commit message of course, explain the purpose, how it works, limitations ;)

and for the final PR, mentions which bug it fixes: 1168, possibly 1260, etc.

**Comment by alban** [comment]
Limitations:
- [ ] IPv6
- [ ] add yours here...

File: docker/tcpv4tracer.py [hunk 1]
```diff
@@ -0,0 +1,291 @@
+#!/usr/bin/python
+#
+# tcpv4tracer   Trace TCP IPv4 connections.
+#               For Linux, uses BCC, eBPF. Embedded C.
+#
```
Comment (alban) [inline docker/tcpv4tracer.py:5]: add a reference of similar tools on the bcc repo: https://github.com/iovisor/bcc/pull/762

File: probe/endpoint/ebpf/main.go [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+import (
+	"fmt"
+	"os"
+	"time"
+
+	"github.com/weaveworks/scope/probe/endpoint"
+)
+
+func main() {
+	tr := endpoint.NewEbpfTracker("/home/asymmetric/code/kinvolk/bcc/examples/tracing/tcpv4tracer.py")
```
Comment (alban) [inline probe/endpoint/ebpf/main.go:12]: hardcoded path.

This file should be converted to Go tests and add some real tests.

File: docker/tcpv4tracer.py [hunk 3]
```diff
@@ -0,0 +1,291 @@
...
+from __future__ import print_function
+from bcc import BPF
+
+import argparse
+import ctypes
+
+parser = argparse.ArgumentParser(
+    description="Trace TCP IPv4 connections",
+    formatter_class=argparse.RawDescriptionHelpFormatter)
+parser.add_argument("-p", "--pid",
```
Comment (alban) [inline docker/tcpv4tracer.py:17]: the `--pid` option is not used in Scope. Do you want to keep it to keep close to the bcc or to remove it?

File: docker/tcpv4tracer.py [hunk 4]
```diff
@@ -0,0 +1,291 @@
...
+#include <net/sock.h>
+#include <net/inet_sock.h>
+#include <net/net_namespace.h>
+#include <bcc/proto.h>
+
+#define TCP_EVENT_TYPE_CONNECT 1
+#define TCP_EVENT_TYPE_ACCEPT  2
+#define TCP_EVENT_TYPE_CLOSE   3
+
+struct tcp_event_t {
```
Comment (alban) [inline docker/tcpv4tracer.py:33]: I see that we diverge from bcc here. So you could remove `--pid` too I guess..

File: docker/tcpv4tracer.py [hunk 5]
```diff
@@ -0,0 +1,291 @@
...
+        }
+
+        if (ret != 0) {
+                // failed to send SYNC packet, may not have populated
+                // socket __sk_common.{skc_rcv_saddr, ...}
+                connectsock.delete(&pid);
+                return 0;
+        }
+
+
```
Comment (alban) [inline docker/tcpv4tracer.py:77]: double empty line

File: probe/endpoint/ebpf.go [hunk 6]
```diff
@@ -0,0 +1,226 @@
...
+type nilTracker struct{}
+
+func (n nilTracker) handleFlow(_ string, _ fourTuple, _ int, _ string) {}
+func (n nilTracker) hasDied() bool                                     { return true }
+func (n nilTracker) run()                                              {}
+func (n nilTracker) walkFlows(f func(ebpfConnection))                  {}
+func (n nilTracker) initialize()                                       {}
+func (n nilTracker) isInitialized() bool                               { return false }
+
+// EbpfTracker contains the list of eBPF events, and the eBPF script's command
```
Comment (alban) [inline probe/endpoint/ebpf.go]: > the list of eBPF events

It is no longer the list of eBPF events but the set of IPv4 TCP connections.

> the eBPF script's command

That comment comes from when the script path was in the struct. It is not true anymore.

File: docker/tcpv4tracer.py (hunk 1 above)
Comment (asymmetric) [reply docker/tcpv4tracer.py:5]: What kind of reference do you mean exactly? BTW this whole file, including this section, is going to change with @iaguis's new tracer.

File: probe/endpoint/ebpf.go [hunk 7]
```diff
@@ -0,0 +1,226 @@
...
+			log.Errorf("EbpfTracker error: %v", err)
+		}
+
+		t.Lock()
+		t.dead = true
+		t.Unlock()
+	}()
+
+	reader := bufio.NewReader(stdout)
+	// skip fist line
```
Comment (alban) [inline probe/endpoint/ebpf.go]: s/fist/first

either explain why you skip the first line or just remove that, add a --no-header option to the python script so that it does not print the header.

File: probe/endpoint/four_tuple.go [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+	fromAddr, toAddr string
+	fromPort, toPort uint16
+}
+
+func (t fourTuple) String() string {
+	return fmt.Sprintf("%s:%d-%s:%d", t.fromAddr, t.fromPort, t.toAddr, t.toPort)
+}
+
+// key is a sortable direction-independent key for tuples, used to look up a
+// fourTuple, when you are unsure of it's direction.
```
Comment (alban) [inline probe/endpoint/four_tuple.go]: s/it's/its

File: probe/endpoint/ebpf.go [hunk 9]
```diff
@@ -0,0 +1,226 @@
...
+
+// EbpfTracker contains the list of eBPF events, and the eBPF script's command
+type EbpfTracker struct {
+	sync.Mutex
+	cmd *exec.Cmd
+
+	initialized bool
+	dead        bool
+
+	activeFlows   map[string]ebpfConnection
```
Comment (asymmetric) [inline probe/endpoint/ebpf.go]: @alban I get the point of trying to mirror the `conntrack.go` module, but would it make more sense to just say `activeConnections`, `walkConnections`,... ?

File: probe/process/walker.go [hunk 10]
```diff
@@ -20,13 +20,14 @@ type Walker interface {
 	Walk(func(Process, Process)) error
 }
 
-// CachingWalker is a walker than caches a copy of the output from another
-// Walker, and then allows other concurrent readers to Walk that copy.
+// CachingWalker is a walker that wraps a platform-specific walker (source),
+// triggers it at every Tick() and caches a copy of the output, to allow
+// other concurrent readers to Walk that copy.
```
Comment (alban) [inline probe/process/walker.go]: The changes in this file belong to another PR: they are unrelated to the ebpf work.

File: probe/endpoint/ebpf.go [hunk 11]
```diff
@@ -0,0 +1,226 @@
...
+	bufferedFlows []ebpfConnection
```
Comment (asymmetric) [inline probe/endpoint/ebpf.go]: I would also call this `closedConnections`, much more descriptive.

File: prog/probe.go [hunk 12]
```diff
@@ -140,8 +140,14 @@ func probeMain(flags probeFlags, targets []appclient.Target) {
 	var scanner procspy.ConnectionScanner
 	if flags.procEnabled {
 		processCache = process.NewCachingWalker(process.NewWalker(flags.procRoot))
-		scanner = procspy.NewConnectionScanner(processCache)
+		processCache.Tick()
 		p.AddTicker(processCache)
+		// if eBPF tracking is enabled, scan /proc synchronously, and just once
```
Comment (alban) [inline prog/probe.go:145]: Note that if eBPF tracking is enabled but does not work (old kernel, or no kernel header or whatever), we need to fall back correctly to the old connection scanner (the asynchronous one)...

File: probe/endpoint/ebpf.go (hunk 11 above)
Comment (asymmetric) [reply probe/endpoint/ebpf.go]: @alban, what do you think?

File: prog/probe.go [hunk 13]
```diff
@@ -158,6 +164,7 @@ func probeMain(flags probeFlags, targets []appclient.Target) {
 		SpyProcs:     flags.spyProcs,
 		UseConntrack: flags.useConntrack,
 		WalkProc:     flags.procEnabled,
+		EbpfEnabled:  flags.ebpfEnabled,
```
Comment (alban) [inline prog/probe.go]: we could track several things with ebpf: processes, connection, ios...

but this PR is only about tracking connections. We should have a better name for the option than a generic "ebpf", we should mention "connections" somewhere.

Ditto for the "EbpfTracker"

File: docker/tcpv4tracer.py (hunk 1 above)
Comment (alban) [reply docker/tcpv4tracer.py:5]: Something like:

```
# tcpv4tracer is similar to https://github.com/iovisor/bcc/pull/762
```

File: probe/endpoint/ebpf.go (hunk 9 above)
Comment (alban) [reply probe/endpoint/ebpf.go]: sounds ok to me.

File: probe/endpoint/ebpf.go (hunk 11 above)
Comment (alban) [reply probe/endpoint/ebpf.go]: ok

Can you add a comment saying that they are recently closed connections and they are not staying forever here but only until the next walkConnections

File: probe/endpoint/four_tuple.go [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+	sort.Strings(key)
+	return strings.Join(key, " ")
+}
+
+// reverse flips the direction of the tuple
+func (t *fourTuple) reverse() {
+	t.fromAddr, t.fromPort, t.toAddr, t.toPort = t.toAddr, t.toPort, t.fromAddr, t.fromPort
+}
+
+// reverse flips the direction of a tuple, without side effects
```
Comment (asymmetric) [inline probe/endpoint/four_tuple.go:35]: @alban do you think it makes sense to have both the function and the method?

File: probe/endpoint/ebpf.go [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+
+func (t *EbpfTracker) hasDied() bool {
+	t.Lock()
+	defer t.Unlock()
+
+	return t.dead
+}
+
+func (t *EbpfTracker) initialize() {
+	t.initialized = true
```
Comment (asymmetric) [inline probe/endpoint/ebpf.go:224]: @alban do you think it's good practice to wrap this in a mutex?

File: probe/endpoint/four_tuple.go (hunk 14 above)
Comment (alban) [reply probe/endpoint/four_tuple.go:35]: if they are used, yes. One is more demanding on the GC than the other.

File: probe/endpoint/ebpf.go (hunk 15 above)
Comment (alban) [reply probe/endpoint/ebpf.go:224]: this does not seem necessary: t.initialized is only read & written from the reporter thread and not from the ebpf tracker "run()" thread.

File: probe/endpoint/four_tuple.go (hunk 14 above)
Comment (asymmetric) [reply probe/endpoint/four_tuple.go:35]: Right, but I could use just the method (which I suppose is less demanding on the GC). I only created the function because I think it looks better :)

**Comment by alban** [comment]
Can you add in the PR the list of tests you did:
- :white_check_mark: get initial connections
- :white_check_mark: get connections from NAT (tested with Kubernetes Services)
- :x: falling back to the old parsing method when ebpf does not work at all (old kernel)
- :x: falling back to the old parsing method when ebpf stops working (python script crashed)
- :x: IPv6
- :question: connection between 2 containers using the same network namespace
- :white_check_mark: displaying connection in the container view when both the source and the destination are pods with multiple containers

**Comment by alepuccetti** [comment]
We are making the official one on scope, don't need this anymore
